server/src/main/java/org/elasticsearch/rest/action/logs/RestLogsAction.java

```diff
@@ -0,0 +1,257 @@
+        XContent xContent = request.getXContentType().xContent();
+        BytesReference content = request.content();
+        byte separator = xContent.streamSeparator();
+        for (int offset = 0, line = 0, endOfEvent; offset < content.length() - 1; offset = endOfEvent + 1, line++) {
+            endOfEvent = content.indexOf(separator, offset);
+            if (endOfEvent == -1) {
+                endOfEvent = content.length();
+            }
+            try (XContentParser parser = xContent.createParser(EMPTY, content.array(), offset, endOfEvent)) {
+                Map<String, Object> event = null;
+                try {
+                    event = parser.map();
+                    if (event.size() == 1 && event.containsKey("_metadata")) {
+                        Map<String, Object> metadata = getMetadata(event);
+                        expandDots(metadata);
+                        if (line == 0) {
+                            MapUtils.recursiveMerge(globalMetadata, metadata);
+                        } else {
+                            localMetadata = metadata;
+                        }
+                        continue;
+                    } else {
+                        expandDots(event);
+                        event = mergeMetadata(event, globalMetadata, localMetadata);
+                        applyDefaultValues(event);
+                    }
+                } catch (Exception e) {
+                    event = mergeMetadata(Map.of(), globalMetadata, localMetadata);
+                    event = createDlqDoc(event, e);
+                }
+                indexRequests.add(Requests.indexRequest(routeToDataStream(event)).opType(DocWriteRequest.OpType.CREATE).source(event));
+            }
+        }
+
+        return channel -> {
+            client.bulk(Requests.bulkRequest().add(indexRequests), new RestActionListener<>(channel) {
+                @Override
+                protected void processResponse(BulkResponse bulkItemResponses) throws Exception {
+                    if (bulkItemResponses.hasFailures() == false) {
+                        sendResponse(channel, RestStatus.ACCEPTED, b -> {});
+                        return;
+                    }
+                    BulkRequest retryBulk = Requests.bulkRequest();
+                    Arrays.stream(bulkItemResponses.getItems()).filter(BulkItemResponse::isFailed).forEach(failedRequest -> {
+                        IndexRequest originalRequest = indexRequests.get(failedRequest.getItemId());
+                        Map<String, Object> doc = originalRequest.sourceAsMap();
+                        BulkItemResponse.Failure failure = failedRequest.getFailure();
+                        if (failure.getStatus() == RestStatus.TOO_MANY_REQUESTS) {
+                            // looks like a transient error; re-try as-is
+                            retryBulk.add(Requests.indexRequest(originalRequest.index()).opType(DocWriteRequest.OpType.CREATE).source(doc));
```

Comment:
There's a risk that re-trying in the target index leads to a persistent error.
I'm inclined to keep the re-try logic simple for now and just always send documents to the DLQ, no matter the reason for ingestion failure. Once we introduce something like a buffer/queue, we could do more clever re-trying on back pressure, such as exponentially backing off the re-tries.

@jpountz, wdyt?

Reply:
That makes sense to me.